src/prefect/environments/storage/s3.py

```diff
@@ -0,0 +1,154 @@
+        self.aws_session_token = aws_session_token
+
+        super().__init__()
+
+    def get_flow(self, flow_location: str) -> "Flow":
+        """
+        Given a flow_location within this Storage object, returns the underlying Flow (if possible).
+
+        Args:
+            - flow_location (str): the location of a flow within this Storage; in this case,
+                a file path where a Flow has been serialized to
+
+        Returns:
+            - Flow: the requested flow
+
+        Raises:
+            - ValueError: if the flow is not contained in this storage
+        """
+        if not flow_location in self.flows.values():
+            raise ValueError("Flow is not contained in this Storage")
+
+        stream = io.BytesIO()
+
+        # Download stream from S3
+        self._boto3_client.download_fileobj(
+            Bucket=self.bucket, Key=flow_location, Fileobj=stream
+        )
+
+        ## prepare data and return
+        stream.seek(0)
+        output = stream.read()
+
+        return cloudpickle.loads(output)
+
+    def add_flow(self, flow: "Flow") -> str:
+        """
+        Method for storing a new flow as bytes in the local filesytem.
+
+        Args:
+            - flow (Flow): a Prefect Flow to add
+
+        Returns:
+            - str: the location of the newly added flow in this Storage object
+
+        Raises:
+            - ValueError: if a flow with the same name is already contained in this storage
+        """
+        if flow.name in self:
+            raise ValueError(
+                'Name conflict: Flow with the name "{}" is already present in this storage.'.format(
+                    flow.name
+                )
+            )
+
+        # Pickle Flow
+        data = cloudpickle.dumps(flow)
+
+        # Write pickled Flow to stream
+        try:
+            stream = io.BytesIO(data)
+        except TypeError:
+            stream = io.BytesIO(data.encode())
+
+        # Create key for Flow that uniquely identifies Flow object in S3
+        key = self.key or "{}/{}".format(
+            slugify(flow.name), slugify(pendulum.now("utc").isoformat())
+        )
+
+        # Upload stream to S3
+        self._boto3_client.upload_fileobj(stream, Bucket=self.bucket, Key=key)
+
+        self.flows[flow.name] = key
+        return key
+
+    def __contains__(self, obj: Any) -> bool:
+        """
+        Method for determining whether an object is contained within this storage.
+        """
+        if not isinstance(obj, str):
+            return False
+        return obj in self.flows
+
+    def build(self) -> "Storage":
+        """
+        Build the Storage object.
+
+        Returns:
+            - Storage: a Storage object that contains information about how and where
+                each flow is stored
+        """
+        return self
+
+    @property
+    def _boto3_client(self):  # type: ignore
+        from boto3 import client as boto3_client
+
+        return boto3_client(
```

Comment:
What happens if all of these are `None`? Does it naturally default to some environment variables and / or some credentials file?

Reply:
Yep! boto3 defaults to the values found in the environment https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#environment-variables